cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb

```diff
@@ -24,19 +24,83 @@
+    annotation_value = gcf_service_account_email
+  when "project-id"
+    annotation_value = "_"
+  when "instance-name"
+    annotation_value = "_"
+  else
+    raise "Incorrect annotation name: '#{annotation_name}', expected: service-account-id|service-account-email|project-id|instance-name"
+  end
+
+  set_annotation_to_resource("authn-gcp/#{annotation_name}", annotation_value)
+end
+
+Given(/^I set "authn-gcp\/(service-account-id|service-account-email)" annotation with value: "([^"]*)" to host "([^"]*)"$/) do |annotation_name, annotation_value, hostname|
+  i_have_a_resource "host", hostname
+  set_annotation_to_resource("authn-gcp/#{annotation_name}", annotation_value)
+end
+
 Given(/^I set all valid GCP annotations to (user|host) "([^"]*)"$/) do |role_type, hostname|
   i_have_a_resource role_type, hostname
 
-  set_annotation_to_resource("authn-gcp/service-account-id", gcp_service_account_id)
-  set_annotation_to_resource("authn-gcp/service-account-email", gcp_service_account_email)
-  set_annotation_to_resource("authn-gcp/project-id", gcp_project_id)
-  set_annotation_to_resource("authn-gcp/instance-name", gcp_instance_name)
+  set_annotation_to_resource("authn-gcp/service-account-id", gce_service_account_id)
+  set_annotation_to_resource("authn-gcp/service-account-email", gce_service_account_email)
+  set_annotation_to_resource("authn-gcp/project-id", gce_project_id)
+  set_annotation_to_resource("authn-gcp/instance-name", gce_instance_name)
+end
+
+Given(/^I set all valid GCF annotations to (user|host) "([^"]*)"$/) do |role_type, hostname|
+  i_have_a_resource role_type, hostname
+
+  set_annotation_to_resource("authn-gcp/service-account-id", gcf_service_account_id)
+  set_annotation_to_resource("authn-gcp/service-account-email", gcf_service_account_email)
+end
+
+Given(/^I set (service-account-id|service-account-email) GCF annotations to (user|host) "([^"]*)"$/) do |annotation_type, role_type, hostname|
+  i_have_a_resource role_type, hostname
+
+  if annotation_type == 'id'
+    set_annotation_to_resource("authn-gcp/service-account-id", gcf_service_account_id)
+  else
+    set_annotation_to_resource("authn-gcp/service-account-email", gcf_service_account_email)
+  end
+end
+
+Given(/^I obtain an? (valid|standard_format|user_audience|invalid_audience|non_existing_host|non_rooted_host|non_existing_account) GCF identity token$/) do |token_type|
+  gcf_identity_access_token(
+    token_type.to_sym
+  )
+end
+
+Given(/I authenticate with authn-gcp using (no|empty|self signed|no kid|obtained|valid) GCF identity token and (non-existing|existing) account/) do |token_state, account|
+  account = account == 'non-existing' ? 'non-existing' : AuthnGcpHelper::ACCOUNT
+
+  token = case token_state
+          when "no"
+            nil
+          when "empty"
+            ""
+          when "self signed"
+            self_signed_token
+          when "no kid"
+            no_kid_self_signed_token
+          else
+            @gcf_identity_token
```

Comment:
can this step be combined with GCE and return `@gcf_identity_token` or `@gce_identity_token` according to the presence of the word `function`?

Reply:
same question for other steps - can we combine some of them to have less code?

Reply:
it's possible but function uses only valid token so I simplified it more.